# Mutual Fund Analytics Capstone
## Day 1 - Data Ingestion

### Objective

The objective of this notebook is to perform data ingestion for the Mutual Fund Analytics project.

Tasks performed:

- Load raw datasets
- Inspect dataset schemas
- Fetch live NAV data from MFAPI
- Validate data availability
- Prepare datasets for ETL processing

This notebook documents the ingestion workflow used in the project.

In [1]:
import pandas as pd
from pathlib import Path

pd.set_option("display.max_columns", None)

## Project Data Sources

The project uses the following datasets:

1. Fund Master Data
2. NAV History
3. Investor Transactions
4. AUM Data
5. SIP Inflows
6. Industry Folio Statistics
7. Benchmark Index Data

Additionally, live NAV data is fetched from MFAPI.

In [2]:
from pathlib import Path

raw_path = Path("../data/raw")

files = []

for file in raw_path.glob("*"):
    files.append(
        {
            "file_name": file.name,
            "size_kb": round(file.stat().st_size / 1024, 2)
        }
    )

pd.DataFrame(files)

,file_name,size_kb
0,01_fund_master.csv,6.66
1,02_nav_history.csv,1190.99
2,03_aum_by_fund_house.csv,3.92
3,04_monthly_sip_inflows.csv,1.63
4,05_category_inflows.csv,3.58
5,06_industry_folio_count.csv,0.80
6,07_scheme_performance.csv,6.44
7,08_investor_transactions.csv,3054.61
8,09_portfolio_holdings.csv,23.35
9,10_benchmark_indices.csv,245.09


## Sample Dataset Inspection

The raw datasets are loaded and inspected to verify schema consistency before ETL processing.

## Raw Data Inventory

The project maintains all source files inside the `data/raw` directory.

The ingestion layer contains:

- Mutual fund master data
- Historical NAV data
- Industry AUM data
- SIP inflow statistics
- Category inflow data
- Folio statistics
- Investor transaction records
- Portfolio holdings
- Benchmark index data
- Live NAV snapshots fetched from MFAPI

The inventory check confirms successful data collection and storage.

In [3]:
fund_master = pd.read_csv("../data/raw/01_fund_master.csv")

print("Shape:", fund_master.shape)

fund_master.head()

Shape: (40, 15)


,amfi_code,fund_house,scheme_name,category,sub_category,plan,launch_date,benchmark,expense_ratio_pct,exit_load_pct,min_sip_amount,min_lumpsum_amount,fund_manager,risk_category,sebi_category_code
0,119551,SBI Mutual Fund,SBI Bluechip Fund - Regular Plan - Growth,Equity,Large Cap,Regular,2006-02-14,NIFTY 100 TRI,1.54,1.0,500,1000,Sohini Andani,Moderate,EC01
1,119552,SBI Mutual Fund,SBI Bluechip Fund - Direct Plan - Growth,Equity,Large Cap,Direct,2013-01-01,NIFTY 100 TRI,0.66,1.0,500,1000,Sohini Andani,Moderate,EC01
2,119598,SBI Mutual Fund,SBI Small Cap Fund - Regular Plan - Growth,Equity,Small Cap,Regular,2009-09-09,BSE 250 SmallCap TRI,1.43,1.0,500,1000,R. Srinivasan,Very High,EC03
3,119599,SBI Mutual Fund,SBI Small Cap Fund - Direct Plan - Growth,Equity,Small Cap,Direct,2013-01-01,BSE 250 SmallCap TRI,0.72,1.0,500,1000,R. Srinivasan,Very High,EC03
4,119120,SBI Mutual Fund,SBI Magnum Gilt Fund - Regular Plan - Growth,Debt,Gilt,Regular,2000-12-30,CRISIL Dynamic Gilt Index,0.77,0.0,500,1000,Dinesh Ahuja,Low,DC02


In [4]:
nav_history = pd.read_csv("../data/raw/02_nav_history.csv")

print("Shape:", nav_history.shape)

nav_history.head()

Shape: (46000, 3)


,amfi_code,date,nav
0,119551,2022-01-03,54.3856
1,119551,2022-01-04,54.3474
2,119551,2022-01-05,54.6869
3,119551,2022-01-06,55.4550
4,119551,2022-01-07,55.3692


In [5]:
investor_txn = pd.read_csv("../data/raw/08_investor_transactions.csv")

print("Shape:", investor_txn.shape)

investor_txn.head()

Shape: (32778, 13)


,investor_id,transaction_date,amfi_code,transaction_type,amount_inr,state,city,city_tier,age_group,gender,annual_income_lakh,payment_mode,kyc_status
0,INV003054,2024-01-01,119092,SIP,1834,Telangana,Hyderabad,T30,56+,Female,77.1,UPI,Verified
1,INV002952,2024-01-01,148567,Redemption,392882,Punjab,Amritsar,B30,18-25,Male,7.1,Cheque,Verified
2,INV003420,2024-01-01,118636,SIP,912,Haryana,Faridabad,B30,36-45,Male,47.2,Mandate,Verified
3,INV003436,2024-01-01,118634,SIP,1102,Maharashtra,Mumbai,T30,36-45,Female,54.4,Cheque,Pending
4,INV004691,2024-01-01,119094,Lumpsum,8682,Delhi,Noida,T30,26-35,Male,14.5,Net Banking,Pending


## Live NAV Data Collection

Live NAV data is collected using the MFAPI service.

Implementation:
- Script: `scripts/live_nav_fetch.py`
- Source: https://api.mfapi.in

The ingestion process fetches NAV information for all selected mutual fund schemes and stores the results in the raw data layer.

In [6]:
live_nav = pd.read_csv("../data/raw/nav_live_all_schemes.csv")

print("Live NAV Records:", len(live_nav))

live_nav.head()

Live NAV Records: 19792


,date,nav,scheme_code,scheme_name,fund_house,scheme_type,scheme_category
0,2013-11-18,13.0894,125497,SBI Small Cap Fund - Direct Plan - Growth,SBI Mutual Fund,Open Ended Schemes,Equity Scheme - Small Cap Fund
1,2013-11-19,13.1068,125497,SBI Small Cap Fund - Direct Plan - Growth,SBI Mutual Fund,Open Ended Schemes,Equity Scheme - Small Cap Fund
2,2013-11-20,12.9549,125497,SBI Small Cap Fund - Direct Plan - Growth,SBI Mutual Fund,Open Ended Schemes,Equity Scheme - Small Cap Fund
3,2013-11-21,12.7957,125497,SBI Small Cap Fund - Direct Plan - Growth,SBI Mutual Fund,Open Ended Schemes,Equity Scheme - Small Cap Fund
4,2013-11-22,12.7775,125497,SBI Small Cap Fund - Direct Plan - Growth,SBI Mutual Fund,Open Ended Schemes,Equity Scheme - Small Cap Fund


In [7]:
print("Fund Master:", fund_master.shape)
print("NAV History:", nav_history.shape)
print("Investor Transactions:", investor_txn.shape)
print("Live NAV Records:", len(live_nav))

Fund Master: (40, 15)
NAV History: (46000, 3)
Investor Transactions: (32778, 13)
Live NAV Records: 19792


In [8]:
summary = pd.DataFrame(
    {
        "Dataset": [
            "Fund Master",
            "NAV History",
            "Investor Transactions",
            "Live NAV"
        ],
        "Records": [
            fund_master.shape[0],
            nav_history.shape[0],
            investor_txn.shape[0],
            len(live_nav)
        ]
    }
)

summary

,Dataset,Records
0,Fund Master,40
1,NAV History,46000
2,Investor Transactions,32778
3,Live NAV,19792


# Data Ingestion Summary

The Day 1 ingestion workflow successfully collected and validated all required datasets.

### Data Volume Overview

- Fund Master Records: 40
- Historical NAV Records: 46,000
- Investor Transaction Records: 32,778
- Live NAV Records: 19,792

### Key Activities Performed

- Loaded raw datasets
- Verified dataset schemas
- Inspected sample records
- Integrated live NAV data using MFAPI
- Prepared data for ETL processing

### Outcome

All datasets were successfully ingested and stored in the raw data layer, enabling downstream cleaning, transformation, analytics, and dashboard development workflows.